Trigger renderModel if an output's model changes

gnestor opened the pull request:
Context: 

[jupyterlab-chart-editor](https://github.com/plotly/jupyterlab-chart-editor) is an IRenderer extension that allows lab users to interactively create charts from data in the notebook and from csv/tsv documents. In a notebook context, if saves the state of the chart editor to the metadata of the output so that the chart is persisted. Given that an outputrea has limited real estate, the “Open new view for output” feature is really helpful. However, when changes were made in the new output view, they were not reflected in the original output. This PR simply observes an output’s model and calls `renderModel` when it changes. I’m hooking this up in `createRenderedMimetype` inside the `OutputArea` class. There may be a better place to do this so I’m open to suggestions…

Before:

![](https://files.gitter.im/jupyterlab/jupyterlab/A5Jl/thumb/imimemodel.gif)

After:

![ezgif-5-c9d63824b4](https://user-images.githubusercontent.com/512354/39588039-c8f840b6-4eaf-11e8-8191-89d9c58c0796.gif)

saulshanabrook commented:
This makes sense from a UX perspective. I don't quite follow the code, since I am not too familiar with how the output area works. Does it work the other way too, so if you update original output will it update the popped out one? And if you have multiple popped out ones will they all update?

gnestor commented:
Thanks for reviewing @saulshanabrook! Yes, different views of an output share the same model, so when that model changes, `renderModel` will be called on each output widget.

jasongrout commented:
Relevant gitter discussion: https://gitter.im/jupyterlab/jupyterlab?at=5aea18bcda323a7c04f0b78a (and the comments following it)

gnestor commented:
I updated this PR to use the existing `OutputAreaModel.stateChanged` signal which was already being emitted but not used. 

`OutputAreaModel.changed` is only emitted in response to a change in the [output list](https://github.com/jupyterlab/jupyterlab/blob/master/packages/outputarea/src/model.ts#L445) whereas `OutputAreaModel.stateChanged` is emitted in response to [`OutputModel.changed`](https://github.com/jupyterlab/jupyterlab/blob/master/packages/outputarea/src/model.ts#L465). 

The only change necessary was to pass the `OutputModel` as an arg vs. `void` with `OutputModel.changed` and `OutputAreaModel.stateChanged` because I couldn't figure out a way to access the model otherwise. On `OutputAreaModel.stateChanged`, the output area gets all of its widgets that are renderers and calls `renderModel` on them.

This also fixes an edge case where `OutputArea._setOutput` expects to get a `Panel` of widgets but gets a `Widget` when there are linked outputs ("Create New View for Output") and something like `IPython.display.update_display` is called on that output area.

vidartf commented on packages/outputarea/src/model.ts:
```diff
@@ -461,15 +460,15 @@ class OutputAreaModel implements IOutputAreaModel {
-  private _onGenericChange(): void {
-    this._stateChanged.emit(void 0);
+  private _onGenericChange(sender: IOutputModel, model: IOutputModel): void {
+    this._stateChanged.emit(model);
```
This could simply emit the `sender` instead of the `model`, right? Then you could avoid the changes to `OutputModel`.

ian-r-rose commented on packages/outputarea/src/model.ts:
```diff
@@ -443,7 +442,7 @@ class OutputAreaModel implements IOutputAreaModel {
     }
     this._changed.emit(args);
-    this._stateChanged.emit(void 0);
+    this._stateChanged.emit(args.newValues[args.newIndex]);
```
I don't think this is correct, as per the [interface](https://github.com/jupyterlab/jupyterlab/blob/master/packages/observables/src/observablelist.ts#L276). The new values are a contiguous list of new values, starting at the index `args.newIndex`. So a general change to the list of outputs has to allow for a list of outputs.

In practice, many outputs areas have just a single `IOutputModel`, and this works for that case, but I think it gives the wrong result for other cases.

ian-r-rose commented on packages/outputarea/src/widget.ts:
```diff
@@ -250,6 +251,19 @@ class OutputArea extends Widget {
+  /**
+   * Follow changes on the model state.
+   */
+  protected onModelStateChanged(sender: IOutputAreaModel, model: IOutputModel): void {
```
Perhaps rather than changing the signal type, we can just access `this.model` here, and not worry about more granular changes to the output model.

gnestor replied:
Ya, my original PR simply did the following at https://github.com/jupyterlab/jupyterlab/blob/master/packages/outputarea/src/widget.ts#L404:

```js
model.stateChanged.connect(() => {
  widget.renderModel(model);
})
```

This was simpler but also seemed more like a hack than this approach. If you have a better idea of how to accomplish this (update an output in response to a change in its model vs. just when a cell is re-run), then feel free to make changes to this branch.

gnestor replied:
Another thing is that `this.model` here is `IOutputAreaModel` but we need `IOutputModel` to call `widget.rendeModel` with...

ian-r-rose commented on packages/outputarea/src/widget.ts:
```diff
@@ -250,6 +251,19 @@ class OutputArea extends Widget {
+  protected onModelStateChanged(sender: IOutputAreaModel, model: IOutputModel): void {
+    let layout = this.layout as PanelLayout;
+    layout.widgets.forEach((panel: Panel) => {
+      let renderer = (panel.widgets ? panel.widgets[1] : panel) as IRenderMime.IRenderer;
```
What is the purpose of this check, here and below?

gnestor replied:
If you create a new view for an output and then call an `IPython.display.update_display` on that output, it will throw an error because for some reason, that _output panel_ doesn't have any `widgets` property.

ellisonbg commented:
@afshin can you help review this one?

gnestor commented:
I simplified this one a lot. Now it supports a output lists and doesn't change any interfaces 👍

To user test:

* Pause the debugger in a `renderModel` function to get a global handle on the model (e.g. `_model = model`)
* Create a linked output ("Create New View for Output")
* Update the model: `_model.setData({ data: { 'text/plain': 'test' } })`
* Both views of the output should update

afshin commented on packages/outputarea/src/model.ts:
```diff
@@ -429,7 +429,6 @@ class OutputAreaModel implements IOutputAreaModel {
     let factory = this.contentFactory;
     let item = factory.createOutputModel(options);
     item.changed.connect(this._onGenericChange, this);
-    item.changed.connect(this._onGenericChange, this);
```
Ha! Nice spot.

afshin commented on packages/outputarea/src/widget.ts:
```diff
@@ -250,6 +251,16 @@ class OutputArea extends Widget {
+  protected onStateChanged(sender: IOutputAreaModel): void {
+    for (let i = 0; i < this.model.length; i++) {
+      this._setOutput(i, this.model.get(i));
+      this.outputLengthChanged.emit(this.model.length);
```
Should this signal be emitted *inside* the loop?
